### PR TITLE
nixos/acme: lego run when account is missing

### LINKED ISSUE #101445: ACME fails with JWS verification error
**Describe the bug**

When upgrading from 20.03 to 20.09, acme started failing for 5 of my domains (but not all of them).

**To Reproduce**
1. systemctl start acme-foo.example.org.service

```
acme-foo.example.com-start[23185]: 2020/10/23 08:32:54 [INFO] [foo.example.com] acme: Obtaining bundled SAN certificate
acme-foo.example.com-start[23185]: 2020/10/23 08:32:54 Could not obtain certificates:
acme-foo.example.com-start[23185]:         acme: error: 400 :: POST :: https://acme-v02.api.letsencrypt.org/acme/new-order :: urn:ietf:params:acme:error:malformed :: JWS verification error, url:
systemd[1]: acme-foo.example.com.service: Main process exited, code=exited, status=1/FAILURE
systemd[1]: acme-foo.example.com.service: Failed with result 'exit-code'.
systemd[1]: Failed to start Renew ACME certificate for foo.example.com.
systemd[1]: acme-foo.example.com.service: Consumed 371ms CPU time, received 5.8K IP traffic, sent 2.2K IP traffic.

```

**Expected behavior**
No failure

**Additional context**

**Notify maintainers**
cc @NixOS/acme 

**Metadata**
 - system: `"x86_64-linux"`
 - host os: `Linux 5.8.14, NixOS, 20.09beta1083.51aaa3fa1b6 (Nightingale)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.7`
 - channels(root): `"nixos-20.09beta1346.05334ad7852, nixos-unstable-21.03pre246543.24c9b05ac53"`
 - channels(symphorien): `"home-manager-20.09"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module: acme
```

### PR DESCRIPTION
**Pull request by m1cr0man**
###### Motivation for this change

Closes #101445

It is possible that the wrong lego command will be issued if no account credentials exist. This can happen if you follow the new instructions to resolve JWS verification errors.

Annoyingly, I did not include the `email` option in the hash for the accounts directory. If I were to add it now, I would trigger a full certificate renewal + account creation for every certificated generated from this module :wink: Thus I will apply Irish "It'll be grand" logic and not change it now.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [X] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [X] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [X] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by m1cr0man** [comment]
Ping ye olde @NixOS/acme

**Comment by andir** [comment]
I wonder if we should use stateVersion to decide if we should use the "compatible" mode or the "proper" mode to handle this. IMHO we should carry a proper solution in the source tree.

**Comment by m1cr0man** [comment]
That sounds like a good solution. I'll add the conditional logic to cover that today.

File: nixos/modules/security/acme.nix
Comment (andir) [inline nixos/modules/security/acme.nix:110]: This looks good. I just had a thought that might also apply in general in these stateVersion checks.

Should we, at some point, throw a warning that a user should migrate (at their pace within a release) so we can realistically remove backwards compatibility at some point?

**Comment by vincentbernat** [comment]
I have hit the bug fixed by this PR. Any chance it gets merged?

**Review by arianvp — APPROVED** [review]

**Comment by m1cr0man** [comment]
Doh! I totally forgot this PR existed. I've duplicated some of this work in #106857 . Can we get this merged so I can deal with the merge conflicts?

File: nixos/modules/security/acme.nix
Comment (flokli) [reply nixos/modules/security/acme.nix:110]: @m1cr0man any opinions on that?

Comment (m1cr0man) [reply nixos/modules/security/acme.nix:110]: So in the newer PR, I actually done away with this logic. This was written under the assumption that there was still rate limit issues with multiple certs being generated, but that's not the case now.

Comment (m1cr0man) [reply nixos/modules/security/acme.nix:110]: Oh also worth noting I added a note on this matter to the [release notes](https://github.com/NixOS/nixpkgs/pull/106857/files#diff-e5a39d8376b580ef7bb2eb1a17900ed609b638aa453fb04f05569bfff04a0173R263).

Comment (flokli) [reply nixos/modules/security/acme.nix:110]: So should this PR be changed before we merge it as requested in https://github.com/NixOS/nixpkgs/pull/101482#issuecomment-745576658? Or do you want to close this one and incorporate it in the other?

Comment (m1cr0man) [reply nixos/modules/security/acme.nix:110]: I will make the necessary changes to this PR so it can be merged. I worry it might take too long for the other PR to be merged.

**Comment by symphorien** [comment]
Do the maintainers think this is fit for a backport to 20.09 ?

**Comment by m1cr0man** [comment]
I was going to propose the back port waits until #106857 is ready, as this PR alone does not fix account creation race conditions. However, it could be argued that has always been an issue so there's no harm in back porting it now. I'll let someone else cast the deciding vote.

**Comment by nh2** [comment]
> However, it could be argued that has always been an issue so there's no harm in back porting it now. I'll let someone else cast the deciding vote.

@m1cr0man Please do backport it, I just found this to block me in the middle of a server migration, and I had to apply these patches to fix it.

**Comment by m1cr0man** [comment]
Ok, the backport is here: https://github.com/NixOS/nixpkgs/pull/107769/files

**Pull request merged**
